### PR TITLE
Ssh scp

### PR DESCRIPTION
**Pull request by arturrez**

### CODE REVIEW

**Comment by arturrez** [comment]
it accepts whatever `scp` command does so  wildcards accepted

**Comment by arturrez** [comment]
latest commit added proper error display
```
❯ ./bin/avalanche node scp  i-0700e01840b65ec39:/home/ubuntu/.avalanchego/logs/main.log ./2/
✗ [i-0700e01840b65ec39] transferring file(s) txt:scp: open local "./2/": No such file or directory
✗ [i-0700e01840b65ec39] transferring file(s) txt:scp: open local "./2/": No such file or directory
 err:exit status 1
Error: failed to scp for node(s) map[aws_node_i-0700e01840b65ec39:exit status 1]
```

And proper destination folders for the case when source it the cluster like 
```
❯ ./bin/avalanche node scp  p0:/home/ubuntu/.avalanchego/logs/main.log ./1
✓ [i-0700e01840b65ec39] transferring file(s)
✓ [i-06164f9106f7dd8d1] transferring file(s)
❯ ls -la ./1/*
./1/p0_i-06164f9106f7dd8d1:
total 32
drwxr-xr-x@ 3 artur  staff     96 21 May 17:00 .
drwxr-xr-x@ 4 artur  staff    128 21 May 16:59 ..
-rw-r--r--@ 1 artur  staff  15172 21 May 17:00 main.log

./1/p0_i-0700e01840b65ec39:
total 32
drwxr-xr-x@ 3 artur  staff     96 21 May 17:00 .
drwxr-xr-x@ 4 artur  staff    128 21 May 16:59 ..
-rw-r--r--@ 1 artur  staff  15176 21 May 17:00 main.log

```

File: pkg/utils/ssh.go [hunk 1]
```diff
@@ -18,10 +19,67 @@ import (
...
+func SplitSCPPath(path string) (string, string) {
+	if !strings.Contains(path, ":") {
+		return "", path
+	}
+	parts := strings.Split(path, ":")
+	return parts[0], parts[1]
+}
+
+// CombineScpPath combines the given host and path into a single item for scp.
+func CombineScpPath(host, path string) string {
```
Comment (felipemadero) [inline pkg/utils/ssh.go:78]: nit: be uniform in the naming, either SCP or Scp

File: pkg/utils/ssh.go [hunk 2]
```diff
@@ -18,10 +19,67 @@ import (
...
+		sourcePath = fmt.Sprintf("%s@%s:%s", constants.AnsibleSSHUser, sourceIP, sourcePath)
+	}
+	if destIP != "" {
+		destPath = fmt.Sprintf("%s@%s:%s", constants.AnsibleSSHUser, destIP, destPath)
+	}
+
+	return fmt.Sprintf("scp %s %s %s", scpParams, sourcePath, destPath), nil
+}
+
+// Command returns an exec.Cmd for the given command line.
```
Comment (felipemadero) [inline pkg/utils/ssh.go:59]: nit: probably this don't belong to ssh.go

File: pkg/utils/ssh.go [hunk 3]
```diff
@@ -18,10 +19,67 @@ import (
...
+	return fmt.Sprintf("ssh %s %s@%s %s", constants.AnsibleSSHShellParams, constants.AnsibleSSHUser, publicIP, certFilePath)
+}
+
+// GetSCPCommandString returns the SCP command string for the given source and destination paths.
+func GetSCPCommandString(certFilePath string, sourceIP, sourcePath string, destIP, destPath string, recursive, withCompression bool) (string, error) {
+	scpParams := constants.AnsibleSSHShellParams + " -B -o LogLevel=Error"
+	if sourceIP == "" && destIP == "" {
+		return "", fmt.Errorf("source or destination should be remote")
+	}
+	if sourceIP == "" && destPath == "" {
```
Comment (felipemadero) [inline pkg/utils/ssh.go:33]: the error msg confuses me: should they be both paths (sourcePath, destPath)? should this be an or(||) op?

File: pkg/utils/ssh.go [hunk 4]
```diff
@@ -18,10 +19,67 @@ import (
...
+		scpParams += " -C"
+	}
+	if certFilePath != "" {
+		scpParams += fmt.Sprintf(" -i %s", certFilePath)
+	}
+	if sourceIP != "" && destIP != "" {
+		scpParams += " -3"
+	}
+	if sourceIP != "" {
+		sourcePath = fmt.Sprintf("%s@%s:%s", constants.AnsibleSSHUser, sourceIP, sourcePath)
```
Comment (felipemadero) [inline pkg/utils/ssh.go:50]: can you use a function for this? probably overload the function you have in this file

File: pkg/utils/ssh.go [hunk 5]
```diff
@@ -18,10 +19,67 @@ import (
...
+	}
+	if destIP != "" {
+		destPath = fmt.Sprintf("%s@%s:%s", constants.AnsibleSSHUser, destIP, destPath)
```
Comment (felipemadero) [inline pkg/utils/ssh.go:53]: same

File: cmd/nodecmd/ssh.go [hunk 6]
```diff
@@ -37,6 +41,9 @@ If no [cmd] is provided for the node, it will open ssh shell there.
 		RunE: sshNode,
 	}
 	cmd.Flags().BoolVar(&isParallel, "parallel", false, "run ssh command on all nodes in parallel")
+	cmd.Flags().BoolVar(&includeMonitor, "with-monitor", false, "include monitoring node for ssh cluster operations")
```
Comment (felipemadero) [inline cmd/nodecmd/ssh.go:44]: the default is being changed here

File: cmd/nodecmd/ssh.go [hunk 7]
```diff
@@ -37,6 +41,9 @@ If no [cmd] is provided for the node, it will open ssh shell there.
...
+	cmd.Flags().BoolVar(&includeLoadTest, "with-loadtest", false, "include loadtest node for ssh cluster operations")
```
Comment (felipemadero) [inline cmd/nodecmd/ssh.go:45]: also here

File: cmd/nodecmd/scp.go [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+		case len(selectedHost) > 2:
+			return nil, ""
+		default:
+			return selectedHost[0], clusterName
+		}
+	}
+	return nil, ""
+}
+
+/*
```
Comment (felipemadero) [inline cmd/nodecmd/scp.go:280]: remove this

File: cmd/nodecmd/scp.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+$ avalanche node scp [cluster1|node1]:/tmp/file.txt /tmp/file.txt
+$ avalanche node scp /tmp/file.txt [cluster1|NodeID-XXXX]:/tmp/file.txt
+$ avalanche node scp node1:/tmp/file.txt NodeID-XXXX:/tmp/file.txt
+`,
+		Args: cobrautils.MinimumNArgs(2),
+		RunE: scpNode,
+	}
+	cmd.Flags().BoolVar(&isRecursive, "recursive", false, "copy directories recursively")
+	cmd.Flags().BoolVar(&withCompression, "compress", false, "use compression for ssh")
+	cmd.Flags().BoolVar(&includeMonitor, "with-monitor", false, "include monitoring node for scp cluster operations")
```
Comment (felipemadero) [inline cmd/nodecmd/scp.go:52]: same here for default. probably it is best to keep true as default (as before)

File: cmd/nodecmd/scp.go [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+// prepareSCPTarget prepares the target for scp command
+func prepareSCPTarget(op ClusterOp, host *models.Host, clusterName string, dest string, isSrc bool) (string, error) {
+	// valid clusterName - is already checked
+	if !strings.Contains(dest, ":") {
+		// destination is local, ready to go
+		return dest, nil
+	}
+	// destination is remote
+	splitDest := strings.Split(dest, ":")
```
Comment (felipemadero) [inline cmd/nodecmd/scp.go:219]: I believe you have a function for this

**Review by felipemadero — COMMENTED** [review]
nits

File: cmd/nodecmd/scp.go (hunk 10 above)
Comment (arturrez) [reply cmd/nodecmd/scp.go:219]: can't find it. can you pls point me ?

File: cmd/nodecmd/ssh.go (hunk 6 above)
Comment (arturrez) [reply cmd/nodecmd/ssh.go:44]: yes, I am aware of it.
monitor and avalanchego nodes are so different. I guess we should mention it in the release notes explicitly

Comment (arturrez) [reply cmd/nodecmd/ssh.go:44]: we should treat avalancego nodes as cluster and service nodes separatly from each other imho

File: cmd/nodecmd/ssh.go (hunk 7 above)
Comment (arturrez) [reply cmd/nodecmd/ssh.go:45]: the same reply

File: pkg/utils/ssh.go (hunk 1 above)
Comment (arturrez) [reply pkg/utils/ssh.go:78]: agree. fixed

File: pkg/utils/ssh.go (hunk 4 above)
Comment (arturrez) [reply pkg/utils/ssh.go:50]: moved

File: cmd/nodecmd/scp.go (hunk 10 above)
Comment (arturrez) [reply cmd/nodecmd/scp.go:219]: ah. thank you. found it and fixed

File: pkg/utils/ssh.go (hunk 2 above)
Comment (arturrez) [reply pkg/utils/ssh.go:59]: moved out. thank you

File: pkg/utils/ssh.go (hunk 4 above)
Comment (arturrez) [reply pkg/utils/ssh.go:50]: done

File: pkg/utils/ssh.go (hunk 5 above)
Comment (arturrez) [reply pkg/utils/ssh.go:53]: done

File: pkg/utils/ssh.go (hunk 3 above)
Comment (arturrez) [reply pkg/utils/ssh.go:33]: good one. my bad. 
fixed

File: cmd/nodecmd/ssh.go (hunk 6 above)
Comment (felipemadero) [reply cmd/nodecmd/ssh.go:44]: What do you think @sukantoraymond

**Review by felipemadero — APPROVED** [review]

File: pkg/utils/ssh.go (hunk 1 above)
Comment (arturrez) [reply pkg/utils/ssh.go:76]: ah. you are correct. fixed now finally with SCP to be consistent with ssh

Comment (arturrez) [reply pkg/utils/ssh.go:76]: I mean SSH everywhere

**Pull request merged**